Documentation: Improve BoxGeometry page

rexdk opened the pull request:
To clarify and correct the BoxGeometry documentation

See #15115

rexdk committed:
Update BoxGeometry.html

To clarify and correct BoxGeometry documentation

rexdk committed:
Update BoxGeometry.html

Mugen87 commented:
https://raw.githack.com/rexdk/three.js/patch-1/docs/index.html#api/en/geometries/BoxGeometry

Mugen87 commented on docs/api/en/geometries/BoxGeometry.html:
```diff
@@ -44,12 +44,13 @@ <h2>Constructor</h2>
+		widthSegments — Number of segmented rectangular faces along the width of the sides. Optional; defaults to 1.<br />
+		heightSegments — Number of segmented rectangular faces along the height of the sides. Optional; defaults to 1.<br />
+		depthSegments — Number of segmented rectangular faces along the depth of the sides. Optional; defaults to 1.<br />
+		Each rectangular face comprises 2 triangular faces.
```
This sentence is confusing. If the segments parameter are greater 1, a single side of the cube consists of more than two triangles: https://jsfiddle.net/f2Lommf5/15451/

Mugen87 commented on docs/api/en/geometries/BoxGeometry.html:
```diff
@@ -59,13 +60,14 @@ <h3>[property:Object parameters]</h3>
+		cube.geometry.parameters; // as above
+		cube.geometry.parameters.width; // === 1
+		cube.geometry.parameters.widthSegments; // === undefined.
+		cube.geometry.vertices.length; // === 8, the number of vertices of the cube; vertices is a property inherited from [page:Geometry]
```
This line can be removed since it is `Geometry` specific.

Mugen87 commented:
You also have to change the API page for `BoxBufferGeometry`.

https://raw.githack.com/rexdk/three.js/patch-1/docs/index.html#api/en/geometries/BoxBufferGeometry

rexdk commented:
Thanks.
1. I'll remove the " ..2 triangular.."  line - if it seems confusing, it is confusing. It is accurate I think, but I did wonder about it.

2. The point about mentioning the inheritance from Geometry is that nowhere in the docs is the concept of inheritance and sub-classes explained. And though it probably seems very easy and natural to a professional js developer, it is quite a hard concept to understand from the code when you first come to it - I certainly found it hard to grasp what the code was doing, not being well versed in javascript. (I've just checked : the W3schools javascript tutorial doesn't appear to mention the sub-class mechanism used in three.js for example, though all the individual bits are there.)

A better approach would be to explain somewhere, probably in the Getting Started section, how it works in simple terms - it's not necessary to explain how, just that some classes inherit properties and methods from other classes, and that that is indicated by the base class and an arrow at the top of the page. But I thought that was too ambitious for my first effort. I'll take your advice - remove the line? Explain about inheritance somewhere? 

3. Yes, I know about BoxBufferGeometry. And LineSegments. And EdgesGeometry.. etc (different issues). Being new, I'm going to spot the things that aren't clear much easier than people who work with it frequently and know it well :)

rexdk committed:
Update BoxGeometry.html

Mugen87 commented:
> remove the line? Explain about inheritance somewhere?

Please remove the line. And we should not to explain inheritance elsewhere in the docs. They are not intended as a guide for object-oriented programming.

rexdk committed:
Update BoxGeometry.html

removed inheritance example as requested

rexdk commented:
OK, removed that line, and I've changed BoxBufferGeometry.html too.

My take on explaining how eg BoxGeometry inherits properties from Geometry is that a simple explanation would be enough for many less experienced users, whereas without an explanation the documentation seems to be aiming at javascript experts. But it's clearly not my decision to make, and that is fine.

looeee commented:
> for many less experienced users

Inheritance should be explained in the first few classes of any object oriented programming course. It's not the purpose of the docs to explain how OOP works, or how JavaScript works. 

We should make it clear that inheritance is happening. To my mind we already do that sufficiently though, with the arrows at the top of each docs page. I'm pretty sure that's a standard way of showing inheritance in API documentation. Feel free to check other library's docs though, and if you find some that you think explain it more clearly, link them here

rexdk commented:
Don't want to fight my corner too strongly here, because it is your project, your decision, not mine. I myself do understand classes and inheritance reasonably from other languages, but did not pick up from the docs that this was happening here, partly because my js understanding is very rusty and partly because the mechanism is fairly advanced in js. For example https://javascript.info/ seems to cover this in chapter 7.7 - that is really near the end, and many people might not get that far! (yet could easily use three.js).

I agree that the arrow is the standard way to indicate inheritance, though it did not say it to me in a js project.

I see that the docs for Line currently have:
"Properties
See the base Object3D class for common properties"
It's simple, short and to my mind helpful and adequate.

But I really don't want to make this a big thing, and respect your views. Just tell me no if that's what you think :)

looeee commented:
> I see that the docs for Line currently have:
"Properties
See the base Object3D class for common properties"
It's simple, short and to my mind helpful and adequate.

Actually, I agree, it would be reasonable to add this note to every relevant docs page. It's quite a bit of work though which is probably why it's only on a few pages so far.

rexdk commented:
A further advantage of including a "See the base xxx class for common properties" line is that a user can quickly get to the inherited properties docs by clicking the base class link.

I'll wait to see if there are further views, then change these files if there is a consensus or no further disagreement.

looeee commented:
> I'll wait to see if there are further views

This is already on quite a few pages, so I don't think there's any need to wait. However, I think further discussion about this should happen in a separate PR. Perhaps you could open one with this added to one or two pages, and we can continue the discussion there?

rexdk committed:
Update BoxGeometry.html

1. added link to base Geometry class for common properties/methods
2. changed description to indicate the cuboid is centred on the origin at creation

rexdk commented:
OK - I've added the base class links now.

I'll open another PR - but I'm not absolutely sure how... The options seem to be
1. Commit directly to the patch-1 branch.
2. Create a new branch for this commit and start a pull request.
so presumably 2. is what I should do? Sorry to be cautious, but I don't want to mess up and start new branches if that's wrong!

mrdoob commented on docs/api/en/geometries/BoxGeometry.html:
```diff
@@ -44,30 +44,36 @@ <h2>Constructor</h2>
 		</p>
 
 		<h2>Properties</h2>
+		<p>See the base [page:Geometry] class for common properties.</p>
```
It'd be better to leave these for a different PR. Currently these additional changes are blocking this PR.

rexdk commented:
OK. I see that a change has been flagged by Mugen87, and I think I have done that change. I'm not sure if that is the hold up. I see there is a button to "Resolve Conversation" and I've now clicked that. I'm not sure if that is for me to do and I'm not sure if there anything else I need to do? (sorry to be ignorant)

looeee commented:
@rexdk just remove this line:

```
<p>See the base [page:Geometry] class for common properties.</p>
```

We'll hopefully add something like this to all the geometry pages, but we'll do it separately to this pull request.

rexdk commented:
I've removed that line, and the similar link under methods. 
I did modify all the other geometry pages to add the links, under a different  branch, called 
docs-add-links-to-base-class. I then created a new pull request as I thought, but I now cannot see whether that pull request still exists.

rexdk committed:
Update BoxGeometry.html

rexdk commented:
Goodness knows what has happened. But I have changed it again.

looeee commented:
OK, looks good now! 👍

mrdoob commented:
Thanks!
